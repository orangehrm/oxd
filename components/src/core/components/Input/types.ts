export interface OutputFile extends Pick<File, 'name' | 'type' | 'size'> {
  base64?: string;
}

export const LEFT = 'left';
export const RIGHT = 'right';
export const TOP = 'top';
export const BOTTOM = 'bottom';
export const LABEL_POSITIONS = [LEFT, RIGHT];
export const DROPDOWN_POSITIONS = [TOP, BOTTOM, LEFT, RIGHT];
export type Position = typeof LEFT | typeof RIGHT | typeof TOP | typeof BOTTOM;

export interface Option {
  id: number;
  label: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [key: string]: any;
}

export type DropdownOptions = Array<Option>;
