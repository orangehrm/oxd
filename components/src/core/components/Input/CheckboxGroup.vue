<script lang="ts">
import {defineComponent, h} from 'vue';
import CheckboxInput from '@orangehrm/oxd/core/components/Input/CheckboxInput.vue';
import InputGroup from '@orangehrm/oxd/core/components/InputField/InputGroup.vue';
import useTranslate from '../../../composables/useTranslate';

export interface State {
  focused: boolean;
}

export interface Options {
  id: number;
  label: string;
  disabled?: boolean;
  style?: Record<string, any>;
}

export default defineComponent({
  name: 'oxd-checkbox-group',
  components: {
    CheckboxInput,
    InputGroup,
  },

  emits: ['update:modelValue', 'blur', 'focus', 'change'],

  props: {
    id: {
      type: String,
      default: 'check-box-group-id',
    },
    class: {
      type: String,
      default: 'check-box-column',
    },
    style: {
      type: Object,
    },
    disabled: {
      type: Boolean,
    },
    options: {
      type: Array,
    },
    modelValue: {
      type: Array,
      default: () => [],
    },
  },

  render() {
    const inputId = this.id == '' ? 'check-box-group-id' : this.id;
    const inputClass = this.class == '' ? 'check-box-column' : this.class;
    const {$t} = useTranslate();
    return h(
      InputGroup,
      {
        classes: {
          wrapper: inputClass,
        },
        id: inputId,
        style: this.style,
      },
      [
        this.options.map((option: Options, index: number) => {
          return h(CheckboxInput, {
            id: inputId + '_' + option.id,
            value: option.id,
            style: option.style,
            autofocus: this.$attrs.autofocus && index === 0 ? true : false,
            optionLabel: $t(option.label),
            modelValue: this.modelValue,
            disabled: this.disabled ? 'true' : option.disabled,
            onFocus: () => {
              this.$emit('focus', true);
            },
            onBlur: () => {
              this.$emit('blur', true);
            },
            onClick: () => {
              this.$emit('click', true);
            },
            onChange: () => {
              let idArray: number[];
              if (this.modelValue.includes(option.id)) {
                idArray = [...this.modelValue].filter(
                  value => value !== option.id,
                );
              } else {
                idArray = [...this.modelValue, option.id];
              }
              this.$emit('update:modelValue', idArray.sort());
              this.$emit('change', true);
            },
          });
        }),
      ],
    );
  },
});
</script>

<style lang="scss" scoped>
:deep(.check-box-row) {
  display: flex;
  gap: 2rem;
}

:deep(.check-box-column) {
  display: flex;
  flex-direction: column;
  gap: 1rem;
}
</style>
