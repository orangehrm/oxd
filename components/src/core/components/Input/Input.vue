<template>
<div :class="outerClasses" class="input-outer-wrapper">
  <div v-if="imageIcon" class="input-text-field-icon"> 
    <img :src="imageIcon" />
  </div>
  <oxd-divider v-if="imageIcon" :orientation="'vertical'" :style="{ height:'100%' }"/>
  <input
    :class="classes"
    :style="style"
    :value="modelValue"
    @focus="onFocus"
    @blur="onBlur"
    @input="onInput"
    v-bind="$attrs"
  />
</div>
</template>

<script lang="ts">
import {defineComponent} from 'vue';
import Divider from '@orangehrm/oxd/core/components/Divider/Divider.vue';

export interface State {
  focused: boolean;
}

export default defineComponent({
  name: 'oxd-input',

  components :{
    'oxd-divider' : Divider
  },

  props: {
    // https://v3.vuejs.org/guide/migration/v-model.html#overview
    modelValue: {},
    style: {
      type: Object,
    },
    hasError: {
      type: Boolean,
      default: false,
    },
    imageIcon: {
      type: String
    }
  },

  data(): State {
    return {
      focused: false,
    };
  },

  computed: {
    classes(): object {
      const isIconExists=this.imageIcon ? '--with-icon' : '';
      return {
        [`oxd-input${isIconExists}`]: true,
        [`oxd-input${isIconExists}--active`]: !this.focused,
        [`oxd-input${isIconExists}--focus`] : this.focused,
        [`oxd-input${isIconExists}--error`] : this.hasError
      };
    },
    outerClasses(): object {
      return this.imageIcon ? {
        'input-outer' : true,
        'input-outer--active': !this.focused,
        'input-outer--focus': this.focused,
        'input-outer--error': this.hasError,
      } : {'input-outer' : false};
    },
  },

  methods: {
    onFocus() {
      this.focused = true;
    },
    onBlur() {
      this.focused = false;
    },
    onInput(e: Event) {
      e.preventDefault();
      this.$emit('update:modelValue', (e.target as HTMLInputElement).value);
      this.$emit('input', e);
    },
  },
});
</script>

<style src="./input.scss" lang="scss" scoped></style>
