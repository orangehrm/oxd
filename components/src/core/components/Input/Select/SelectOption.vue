<template>
  <div @mousedown="onClick" role="option" class="oxd-select-option">
    <slot></slot>
  </div>
</template>

<script lang="ts">
import {defineComponent} from 'vue';

export default defineComponent({
  name: 'oxd-select-option',

  props: {
    disabled: {
      type: Boolean,
      default: false,
    },
  },

  emits: ['select'],

  methods: {
    onClick() {
      if (this.disabled) return;
      this.$emit('select');
    },
  },
});
</script>

<style src="./select-input.scss" lang="scss" scoped></style>
