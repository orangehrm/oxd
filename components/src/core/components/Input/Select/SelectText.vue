<template>
  <div
    :class="classes"
    :style="style"
    :tabindex="tabIndex"
    v-bind="defaultAttrs"
    @focus="onFocus"
    @blur="onBlur"
  >
    <div class="oxd-select-text-input">
      <div class="selected-content">{{ value }}</div>
      <input
        type="text"
        readonly="readonly"
        v-bind="inputAttrs"
        @focus="onFocus"
        @blur="$attrs.onBlur"
      />
    </div>
    <div class="oxd-select-text--after-slot-option">
      <slot name="afterInput"></slot>
    </div>
    <div class="oxd-select-text--after">
      <div :class="dropdownIconClasses">
        <oxd-icon
          v-if="!disabled"
          :class="dropdownIconClasses"
          :size="dropdownIconSize"
          :name="dropdownIcon"
        />
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import {defineComponent} from 'vue';
import Icon from '@orangehrm/oxd/core/components/Icon/Icon.vue';
import eventsMixin from './events-mixin';
import navigationMixin from './navigation-mixin';

export default defineComponent({
  name: 'oxd-select-text',
  inheritAttrs: false,

  components: {
    'oxd-icon': Icon,
  },

  mixins: [navigationMixin, eventsMixin],

  emits: ['icon:clicked'],

  props: {
    dropdownOpened: {
      type: Boolean,
      default: true,
    },
    value: {
      type: String,
    },
    style: {
      type: Object,
    },
    hasError: {
      type: Boolean,
      default: false,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    readonly: {
      type: Boolean,
      default: false,
    },
    downIcon: {
      type: String,
      default: 'caret-down-fill',
    },
    upIcon: {
      type: String,
      default: 'caret-up-fill',
    },
    dropdownIconSize: {
      type: String,
      default: 'medium',
    },
  },

  data() {
    return {
      focused: false,
    };
  },

  computed: {
    classes(): object {
      return {
        'oxd-select-text': true,
        'oxd-select-text--active': !this.focused,
        'oxd-select-text--focus': this.focused,
        'oxd-select-text--error': this.hasError,
        'oxd-select-text--disabled': this.disabled,
        'oxd-select-text--readonly': this.readonly,
      };
    },
    dropdownIcon(): string {
      return this.dropdownOpened ? this.upIcon : this.downIcon;
    },
    dropdownIconClasses(): object {
      return {
        'oxd-select-text--arrow': true,
        '--disabled': this.disabled,
        '--readonly': this.readonly,
      };
    },
    tabIndex(): number {
      return this.disabled ? -1 : 0;
    },
    inputAttrs() {
      const allowed = ['id'];
      return Object.keys(this.$attrs)
        .filter(key => allowed.includes(key))
        .reduce((obj, key) => {
          obj[key] = this.$attrs[key];
          return obj;
        }, {});
    },
    defaultAttrs() {
      const notAllowed = ['id'];
      return Object.keys(this.$attrs)
        .filter(key => !notAllowed.includes(key))
        .reduce((obj, key) => {
          obj[key] = this.$attrs[key];
          return obj;
        }, {});
    },
  },

  methods: {
    onFocus($e: Event) {
      if (this.disabled) {
        $e.stopImmediatePropagation();
        return;
      }
      this.focused = true;
    },
    onBlur() {
      this.focused = false;
    },
  },
});
</script>

<style src="./select-input.scss" lang="scss" scoped></style>
