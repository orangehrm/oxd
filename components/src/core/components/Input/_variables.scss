// This file is part of OrangeHRM Inc
//
// Copyright (C) 2020 onwards OrangeHRM Inc
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see  http://www.gnu.org/licenses

$oxd-input-min-height: 45px !default;
$oxd-input-border--active: $oxd-input-control-border--active !default;
$oxd-input-border--focus: $oxd-input-control-border--focus !default;
$oxd-input-border--error: $oxd-input-control-border--error !default;
$oxd-input-box-shadow--focus: $oxd-input-control-box-shadow--focus !default;
$oxd-input-box-shadow--error: $oxd-input-control-box-shadow--error !default;
$oxd-input-background--disabled: darken(
  $oxd-background-white-shadow-color,
  3%
) !default;

$oxd-file-input-horizontal-margin: 0.25rem !default;
$oxd-file-input-background-color: $oxd-white-color !default;
$oxd-file-button-background-color: $oxd-background-light-gray-color !default;
$oxd-file-button-color: $oxd-interface-gray-darken-1-color !default;
$oxd-file-button-font-size: 0.875rem;
$oxd-file-button-font-weight: 600;
$oxd-file-button-padding-horizontal: 1.5rem;
$oxd-file-button-color--disabled: $oxd-interface-gray-lighten-1-color !default;
$oxd-file-button-background-color--disabled: darken(
  $oxd-white-color,
  1%
) !default;
$oxd-file-button-height: 35px !default;
$oxd-file-button-width: 35px !default;

$oxd-checkbox-label-color: $oxd-input-control-font-color !default;
$oxd-checkbox-label-color--disabled: $oxd-interface-gray-color !default;
$oxd-checkbox-check-font-size: 1rem !default;
$oxd-checkbox-height: 1rem !default;
$oxd-checkbox-width: 1rem !default;
$oxd-checkbox-border: 1px solid $oxd-interface-gray-lighten-1-color !default;
$oxd-checkbox-border-radius: 0.35rem !default;
$oxd-checkbox-label-margin: 0.5em !default;
$oxd-checkbox-checked-background-color: $oxd-primary-one-color !default;
$oxd-checkbox-disabled-background-color: $oxd-interface-gray-lighten-2-color !default;

$oxd-switch-label-color: $oxd-input-control-font-color !default;
$oxd-switch-label-color--disabled: $oxd-interface-gray-color !default;
$oxd-switch-width: 2rem !default;
$oxd-switch-height: calc(#{$oxd-switch-width} / 2);
$oxd-switch-radius: calc(#{$oxd-switch-height} / 2);
$oxd-switch-selector-height: calc(#{$oxd-switch-height} - 2px);
$oxd-switch-selector-width: calc(#{$oxd-switch-height} - 2px);
$oxd-switch-selector-radius: calc(#{$oxd-switch-selector-height} / 2);
$oxd-switch-selector-color: $oxd-white-color !default;
$oxd-switch-label-margin: 0.5em;
$oxd-switch-off-background-color: $oxd-interface-gray-lighten-2-color !default;
$oxd-switch-on-background-color: $oxd-primary-one-color !default;
$oxd-switch-on-disabled-background-color: rgba(
  $oxd-primary-one-color,
  0.5
) !default;
$oxd-switch-off-disabled-background-color: $oxd-interface-gray-lighten-2-color !default;
$oxd-switch-disabled-background-color: $oxd-interface-gray-lighten-2-color !default;

$oxd-radio-label-color: $oxd-input-control-font-color !default;
$oxd-radio-label-color--disabled: $oxd-interface-gray-color !default;
$oxd-radio-height: 1rem !default;
$oxd-radio-width: 1rem !default;
$oxd-radio-label-margin: 0.5em !default;
$oxd-radio-checked-color: $oxd-primary-one-color !default;
$oxd-radio-checked-disabled-color: lighten(
  $oxd-radio-checked-color,
  30%
) !default;

$oxd-dropdown-text-padding: 0 0.25rem 0 0.5rem !default;
$oxd-dropdown-animation-time: 300ms !default;
$oxd-dropdown-max-height: 11.5rem !default;
$oxd-dropdown-loader-height: 1rem !default;
$oxd-dropdown-dropdown-z-index: $oxd-dropdown-z-index !default;
$oxd-dropdown-dropdown-box-shadow: 0px 5px 15px -8px rgba(0, 0, 0, 0.15) !default;
$oxd-dropdown-dropdown-padding: 4px !default;
$oxd-dropdown-dropdown-border: 1px solid $oxd-interface-gray-lighten-2-color !default;
$oxd-dropdown-dropdown-border-radius: 0.65rem !default;
$oxd-dropdown-dropdown-background: $oxd-white-color !default;
$oxd-dropdown-dropdown-background--disabled: darken(
  $oxd-background-white-shadow-color,
  3%
) !default;
$oxd-dropdown-dropdown-margin: 5px !default;
$oxd-dropdown-dropdown-color: $oxd-interface-gray-darken-1-color !default;
$oxd-dropdown-dropdown-slot-color: $oxd-interface-gray-color !default;
$oxd-dropdown-dropdown-max-height: 224px !default;

$oxd-dropdown-option-font-size: $oxd-input-control-font-size !default;
$oxd-dropdown-option-font-family: $oxd-input-control-font-family !default;
$oxd-dropdown-option-font-weight: $oxd-input-control-font-weight !default;
$oxd-dropdown-option-padding: 4px !default;
$oxd-dropdown-option-min-height: 27px !default;
$oxd-dropdown-option-line-height: normal !default;
$oxd-dropdown-option-border-bottom: 1px solid #fbfbfc !default;
$oxd-dropdown-option-border-radius: 0.65rem !default;
$oxd-dropdown-option-color: $oxd-interface-gray-darken-1-color !default;
$oxd-dropdown-option-color--disabled: $oxd-interface-gray-lighten-1-color !default;
$oxd-dropdown-option-color--selected: $oxd-interface-gray-lighten-1-color !default;
$oxd-dropdown-option-background: transparent !default;
$oxd-dropdown-option-background--focused: #f6f7f9 !default;

$oxd-dropdown-chip-margin: 5px !default;
$oxd-dropdown-chip-background-color: #f6f7f9 !default;
$oxd-dropdown-chip-clear-background-color: $oxd-interface-gray-lighten-1-color !default;
$oxd-dropdown-chip-clear-font-color: $oxd-white-color !default;
$oxd-dropdown-chip-border-radius: 50% !default;
$oxd-dropdown-chip-area-padding: 0.5rem 0 !default;

$oxd-dropdown-arrow-background-color: $oxd-background-light-gray-color !default;
$oxd-dropdown-arrow-background-color--disabled: $oxd-background-light-gray-color !default;
$oxd-dropdown-arrow-background-color--readonly: darken(
  $oxd-white-color,
  1%
) !default;
$oxd-dropdown-arrow-color--readonly: $oxd-interface-gray-lighten-1-color !default;
$oxd-dropdown-arrow-width: 35px !default;
$oxd-dropdown-arrow-height: 35px !default;
$oxd-dropdown-arrow-border-radius: 0.65rem !default;

$oxd-time-picker-period-label-width: 60px !default;
$oxd-time-picker-period-label-padding: 0.25rem 0rem !default;
$oxd-time-picker-period-horizontal-padding: 16px !default;
$oxd-time-picker-period-vertical-padding: 8px !default;
$oxd-time-picker-period-font-size: 14px !default;
$oxd-time-picker-period-font-weight: 600 !default;
$oxd-time-picker-period-font-family: $oxd-font-family !default;
$oxd-time-picker-period-font-color: $oxd-white-color !default;
$oxd-time-picker-period-border-radius: 0.5rem !default;
$oxd-time-picker-dropdown-margin: 1.5rem 1rem !default;
$oxd-time-picker-dropdown-width: 250px !default;
$oxd-time-picker-dropdown-input-width: 60px !default;
$oxd-time-picker-dropdown-seperator-padding: 0.5rem !default;

$oxd-date-input-links-margin: 0.5rem !default;
$oxd-date-input-link-padding: 5px 10px !default;
$oxd-date-input-link-border-radius: 0.75rem !default;
$oxd-date-input-link-font-family: $oxd-input-control-font-family !default;
$oxd-date-input-link-font-color: $oxd-input-control-font-color !default;
$oxd-date-input-link-font-weight: 500 !default;
$oxd-date-input-link-font-size: 14px !default;
$oxd-date-input-link-highlight-color: $oxd-primary-one-color !default;
$oxd-date-input-link-background: $oxd-background-light-gray-color !default;
