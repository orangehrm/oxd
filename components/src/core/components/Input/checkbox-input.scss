@import '../../../styles';
@import 'variables';

.oxd-checkbox-wrapper {
  & label {
    cursor: pointer;
    display: flex;
    align-content: center;
    font-family: $oxd-input-control-font-family;
    color: $oxd-checkbox-label-color;
    font-size: $oxd-input-control-font-size;
    font-weight: $oxd-input-control-font-weight;
    width: fit-content;

    &:after {
      content: '';
      display: table;
      clear: both;
    }

    & input[type='checkbox'] {
      opacity: 0;
      position: absolute;
      + .oxd-checkbox-input {
        background-color: $oxd-white-color;
        .oxd-checkbox-input-icon {
          opacity: 0;
        }
      }

      &:checked + .oxd-checkbox-input {
        background-color: $oxd-checkbox-checked-background-color;
        .oxd-checkbox-input-icon {
          opacity: 1;
        }
      }

      &:checked:disabled + .oxd-checkbox-input {
        background-color: $oxd-checkbox-disabled-background-color;
        .oxd-checkbox-input-icon {
          opacity: 1;
        }
      }

      &:disabled + .oxd-checkbox-input {
        background-color: $oxd-checkbox-disabled-background-color;
        .oxd-checkbox-input-icon {
          opacity: 0;
        }
      }
    }

    &.--disabled {
      color: $oxd-checkbox-label-color--disabled;
      cursor: unset;
    }
  }
}

.oxd-checkbox-input {
  position: relative;
  display: flex;
  border-radius: $oxd-checkbox-border-radius;
  width: $oxd-checkbox-width;
  height: $oxd-checkbox-height;

  &.--label-left {
    margin-left: $oxd-checkbox-label-margin;
  }

  &.--label-right {
    margin-right: $oxd-checkbox-label-margin;
  }

  &--active {
    border: $oxd-checkbox-border;
  }

  &--focus {
    border: $oxd-input-border--focus;
    box-shadow: $oxd-input-box-shadow--focus;
    outline: 0;
  }

  &--error {
    border: $oxd-input-border--error;
    box-shadow: $oxd-input-box-shadow--error;
  }

  &-icon {
    color: $oxd-white-color;
    font-size: $oxd-checkbox-check-font-size;
    position: absolute;
    line-height: 1.5;
    align-self: center;
  }
}
