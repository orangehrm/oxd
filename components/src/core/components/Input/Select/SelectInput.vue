<template>
  <div class="oxd-select-wrapper">
    <oxd-select-text
      v-bind="$attrs"
      :value="inputValue"
      :disabled="disabled"
      :dropdownOpened="dropdownOpen"
      @icon:clicked="onOpenDropdown"
      @blur="onBlur"
      @click="onOpenDropdown"
      @keyup.esc="onCloseDropdown"
      @keydown.enter.prevent="onSelectEnter"
      @keydown.down.exact.prevent="onSelectDown"
      @keydown.up.exact.prevent="onSelectUp"
      @keydown="onKeypress"
    >
      <template v-slot:afterInput>
        <slot v-if="modelValue" name="afterSelected" :data="modelValue"></slot>
      </template>
    </oxd-select-text>

    <oxd-select-dropdown
      v-if="dropdownOpen"
      :class="dropdownClasses"
      :loading="loading"
      :empty="computedOptions.length === 0"
    >
      <oxd-select-option
        v-if="showEmptySelector && !hideDropdownDefaultLabel"
        @select="onClear"
      >
        {{ $vt(placeholder) }}
      </oxd-select-option>
      <oxd-select-option
        v-for="(option, i) in computedOptions"
        :key="option.id"
        :class="optionClasses[i]"
        :disabled="option._disabled || option._selected"
        :ref="`option-${i}`"
        @select="onSelect(option)"
      >
        <slot name="option" :data="option"></slot>
        <span v-if="!$slots['option']">{{ $vt(option.label) }}</span>
      </oxd-select-option>
    </oxd-select-dropdown>
  </div>
</template>

<script lang="ts">
import {defineComponent} from 'vue';
import eventsMixin from './events-mixin';
import navigationMixin from './navigation-mixin';
import {TOP, BOTTOM, Option, Position, DROPDOWN_POSITIONS} from '../types';
import SelectText from '@orangehrm/oxd/core/components/Input/Select/SelectText.vue';
import SelectDropdown from '@orangehrm/oxd/core/components/Input/Select/SelectDropdown.vue';
import SelectOption from '@orangehrm/oxd/core/components/Input/Select/SelectOption.vue';
import translateMixin from '../../../../mixins/translate';

export default defineComponent({
  name: 'oxd-select-input',
  inheritAttrs: false,

  components: {
    'oxd-select-text': SelectText,
    'oxd-select-dropdown': SelectDropdown,
    'oxd-select-option': SelectOption,
  },

  mixins: [navigationMixin, eventsMixin, translateMixin],

  emits: [
    'update:modelValue',
    'dropdown:opened',
    'dropdown:closed',
    'dropdown:blur',
    'dropdown:clear',
  ],

  props: {
    modelValue: {
      type: Object,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    hideDropdownDefaultLabel: {
      type: Boolean,
      default: false,
    },
    options: {
      type: Array,
      required: true,
    },
    placeholder: {
      type: String,
      default: '-- Select --',
    },
    dropdownPosition: {
      type: String,
      default: BOTTOM,
      validator: function(value: Position) {
        return DROPDOWN_POSITIONS.indexOf(value) !== -1;
      },
    },
    showEmptySelector: {
      type: Boolean,
      default: true,
    },
  },

  data() {
    return {
      focused: false,
      loading: false,
      dropdownOpen: false,
      searchTerm: null,
    };
  },

  computed: {
    computedOptions(): Option[] {
      return this.options.map((option: Option) => {
        let _selected = false;
        if (this.modelValue?.id === option.id) {
          _selected = true;
        }
        return {...option, _selected};
      });
    },
    dropdownClasses(): object {
      return {
        '--positon-bottom': this.dropdownPosition === BOTTOM,
        '--positon-top': this.dropdownPosition === TOP,
      };
    },
    optionClasses(): object[] {
      return this.computedOptions.map((option: Option, index: number) => {
        return {
          '--disabled': option._disabled,
          '--selected': option._selected,
          '--focused': index === this.pointer,
          [`--indent-${option._indent}`]: option._indent !== undefined,
        };
      });
    },
    selectedItem(): string {
      return this.modelValue?.label
        ? this.$vt(this.modelValue.label)
        : this.hideDropdownDefaultLabel
        ? null
        : this.$vt(this.placeholder);
    },
    inputValue(): string {
      return this.computedOptions[this.pointer]?.label || this.selectedItem;
    },
  },

  watch: {
    pointer(newIndex: number) {
      const option = this.$refs[`option-${newIndex}`];
      if (option?.$el) this.scrollToView(option.$el);
    },
  },
});
</script>

<style src="./select-input.scss" lang="scss" scoped></style>
