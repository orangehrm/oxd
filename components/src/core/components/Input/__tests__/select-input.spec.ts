import {flushPromises, mount, shallowMount} from '@vue/test-utils';
import SelectInput from '@orangehrm/oxd/core/components/Input/Select/SelectInput.vue';
import SelectText from '@orangehrm/oxd/core/components/Input/Select/SelectText.vue';
import SelectOption from '@orangehrm/oxd/core/components/Input/Select/SelectOption.vue';
import {BOTTOM} from '@orangehrm/oxd/core/components/Input/types';

const options = [
  {
    id: 1,
    label: 'HR Admin',
  },
  {
    id: 2,
    label: 'ESS User',
  },
  {
    id: 3,
    label: 'Supervisor',
  },
];

describe('SelectInput.vue', () => {
  it('renders OXD Select Input', () => {
    const wrapper = mount(SelectInput, {
      props: {options},
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('should load options to Select', async () => {
    const wrapper = mount(SelectInput, {
      props: {options},
    });
    wrapper.findComponent(SelectText).trigger('click');
    await wrapper.vm.$nextTick();
    const nodes = wrapper.findAllComponents(SelectOption);
    expect(nodes.length).toBe(4);
  });
  it('should select one option', async () => {
    const wrapper = mount(SelectInput, {
      props: {options},
    });
    wrapper.findComponent(SelectText).trigger('click');
    await wrapper.vm.$nextTick();
    const nodes = wrapper.findAllComponents(SelectOption);
    await nodes[1].trigger('mousedown');
    expect(wrapper.emitted('update:modelValue')).toEqual([
      [
        {
          id: 1,
          label: 'HR Admin',
          _selected: false,
        },
      ],
    ]);
  });
  it('on Focus it should add class "oxd-select-text--focus"', async () => {
    const wrapper = mount(SelectInput, {
      props: {options, readonly: false, disabled: false},
    });
    wrapper.findComponent(SelectText).trigger('focus');
    await wrapper.vm.$nextTick();
    expect(wrapper.find('.oxd-select-text--focus').exists()).toBe(true);
  });

  it('on Focus it should not add class "oxd-select-text--focus" when Select is disabled', async () => {
    const wrapper = mount(SelectInput, {
      props: {options, readonly: false, disabled: true},
    });
    wrapper.findComponent(SelectText).trigger('focus');
    await wrapper.vm.$nextTick();
    expect(wrapper.find('.oxd-select-text--focus').exists()).toBe(false);
  });

  it('on Blur it should add class "oxd-select-text--active"', async () => {
    const wrapper = mount(SelectInput, {
      props: {options, readonly: false, disabled: false},
    });
    wrapper.findComponent(SelectText).trigger('blur');
    await wrapper.vm.$nextTick();
    expect(wrapper.find('.oxd-select-text--active').exists()).toBe(true);
  });

  it('Id filed append to the input field of Select"', async () => {
    const wrapper = mount(SelectInput, {
      props: {id: 'form_select', options},
    });
    wrapper.findComponent(SelectText).trigger('blur');
    await wrapper.vm.$nextTick();
    const Input = wrapper.find('input');
    expect(Input.attributes('id')).toBe('form_select');
  });

  it('Class filed not to append to the input field of Select"', async () => {
    const wrapper = mount(SelectInput, {
      props: {class: 'sampleClass', options},
    });
    wrapper.findComponent(SelectText).trigger('blur');
    await wrapper.vm.$nextTick();
    const Input = wrapper.find('input');
    expect(Input.classes('sampleClass')).not.toBe(true);
  });

  it('should select none if placeholder selected', async () => {
    const wrapper = mount(SelectInput, {
      props: {options},
    });
    wrapper.findComponent(SelectText).trigger('click');
    await wrapper.vm.$nextTick();
    const nodes = wrapper.findAllComponents(SelectOption);
    await nodes[0].trigger('mousedown');
    expect(wrapper.emitted('update:modelValue')).toEqual([[null]]);
  });
  it('should not select already selected option', async () => {
    const wrapper = mount(SelectInput, {
      props: {
        options,
        modelValue: {
          id: 1,
          label: 'HR Admin',
        },
      },
    });
    wrapper.findComponent(SelectText).trigger('click');
    await wrapper.vm.$nextTick();
    const nodes = wrapper.findAllComponents(SelectOption);
    await nodes[1].trigger('mousedown');
    expect(wrapper.emitted('update:modelValue')).toBeFalsy();
  });
  it('should not have placeholder', async () => {
    const wrapper = mount(SelectInput, {
      props: {
        options,
        modelValue: {
          id: 1,
          label: 'HR Admin',
        },
        showEmptySelector: false,
      },
    });
    wrapper.findComponent(SelectText).trigger('click');
    await wrapper.vm.$nextTick();
    const nodes = wrapper.findAllComponents(SelectOption);
    await nodes[0].trigger('mousedown');
    expect(nodes.length).toBe(3);
  });

  it('When place holder is there should have class --with-empty-selector', async () => {
    const wrapper = shallowMount(SelectInput, {
      props: {
        options,
        modelValue: {
          id: 1,
          label: 'HR Admin',
        },
        dropdownPosition: BOTTOM,
        showEmptySelector: true,
        hideDropdownDefaultLabel: false,
      },
    });
    expect(wrapper.vm.dropdownClasses).toStrictEqual({
      '--positon-bottom': true,
      '--positon-top': false,
      '--with-empty-selector': true,
    });
  });

  it('When place holder not there should not have class --with-empty-selector', async () => {
    const wrapper = shallowMount(SelectInput, {
      props: {
        options,
        modelValue: {
          id: 1,
          label: 'HR Admin',
        },
        dropdownPosition: BOTTOM,
        showEmptySelector: false,
        hideDropdownDefaultLabel: false,
      },
    });
    expect(wrapper.vm.dropdownClasses).toStrictEqual({
      '--positon-bottom': true,
      '--positon-top': false,
      '--with-empty-selector': false,
    });
  });

  it('When place holder is hidden should not have class --with-empty-selector', async () => {
    const wrapper = shallowMount(SelectInput, {
      props: {
        options,
        modelValue: {
          id: 1,
          label: 'HR Admin',
        },
        dropdownPosition: BOTTOM,
        showEmptySelector: true,
        hideDropdownDefaultLabel: true,
      },
    });
    expect(wrapper.vm.dropdownClasses).toStrictEqual({
      '--positon-bottom': true,
      '--positon-top': false,
      '--with-empty-selector': false,
    });
  });

  it('when click on icon opens the dropdown', async () => {
    const wrapper = mount(SelectInput, {
      props: {options},
    });
    wrapper.find('.oxd-select-text--after i.oxd-icon').trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('click'));

    wrapper.find('.oxd-select-text--after i.oxd-icon').trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('click'));
  });

  it('when click on input to opens the dropdown', async () => {
    const wrapper = mount(SelectInput, {
      props: {options},
    });
    wrapper.find('.oxd-select-text').trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('click'));

    wrapper.find('.oxd-select-text').trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('click'));
  });
});
