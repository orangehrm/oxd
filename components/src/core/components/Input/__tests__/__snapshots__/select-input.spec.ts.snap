// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectInput.vue renders OXD Select Input 1`] = `
<div class="oxd-select-wrapper">
  <div class="oxd-select-text oxd-select-text--active" tabindex="0">
    <div class="oxd-select-text-input">
      <div class="selected-content">-- Select --</div><input type="text" readonly="">
    </div>
    <div class="oxd-select-text--after-slot-option">- </div>
    <div class="oxd-select-text--after">
      <div class="oxd-select-text--arrow"><i class="oxd-icon oxd-icon--medium bi-caret-down-fill oxd-select-text--arrow"></i></div>
    </div>
  </div>
  <!--v-if-->
</div>
`;
